### PR TITLE
[FLINK-9630] [connector] Kafka09PartitionDiscoverer cause connection …

### PR DESCRIPTION
**Pull request by ubyyj**
…leak on TopicAuthorizationException

**(The sections below can be removed for hotfixes of typos)**

## What is the purpose of the change
Fix the bug that Kafka09PartitionDiscoverer can cause TCP connection leak, if getAllPartitionsForTopics() get a TopicAuthorizationException.

## Brief change log
catch TopicAuthorizationException and close the kafkaConsumer in getAllPartitionsForTopics().

## Verifying this change
This change added tests and can be verified as follows:
 - *Manually verified the change by running job which consumes from an non-exist kafka topic, and verified the # of opened TCP connection and # file handle did not increase of the task manager process, The fix has beening running in our production for weeks now, without problem *

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by ubyyj** [comment]
unrelavent build failure, close and will reopen again, to trigger travis build again

**Comment by ubyyj** [comment]
trigger travis build

**Comment by yanghua** [comment]
@ubyyj you can use this command to trigger Travis rebuild : 

```
git commit --allow-empty -m "XXXX"
```

**Comment by zhangminglei** [comment]
Hi, @ubyyj Thanks for your contribution! It looks good to me.

**Comment by ubyyj** [comment]
@tzulitai is there anything else we need to do before we get this merged?

**Comment by ubyyj** [comment]
@zhangminglei would you please help me get this merged? thanks

**Comment by yanghua** [comment]
To merge PR, you can ping @tillrohrmann and @zentol

**Comment by yanghua** [comment]
hi @ubyyj Till and Chesnay are busy releasing Flink 1.6 and 1.5.2, just wait~

File: flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09PartitionDiscoverer.java [hunk 1]
```diff
@@ -102,4 +106,13 @@ protected void closeConnections() throws Exception {
 			this.kafkaConsumer = null;
 		}
 	}
+
+	private void closeConnectionsQuietly() {
+		try {
+			closeConnections();
```
Comment (walterddr) [inline flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09PartitionDiscoverer.java]: I think you need to invoke the based class `AbstractPartitionDiscoverer.close()` to ensure that the state is also set to closed

File: flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09PartitionDiscoverer.java [hunk 2]
```diff
@@ -81,6 +82,9 @@ protected void initializeConnections() {
 		} catch (org.apache.kafka.common.errors.WakeupException e) {
 			// rethrow our own wakeup exception
 			throw new WakeupException();
+		} catch (TopicAuthorizationException tae) {
+			closeConnectionsQuietly();
```
Comment (walterddr) [inline flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09PartitionDiscoverer.java]:
```suggestion
			try {
			  close();
			} catch (Exception e) {
			  throw new RuntimeException("Error while closing partition discoverer.", e);
			}
```

**Comment by aljoscha** [comment]
The Kafka 0.9 code was removed from Flink by now.

**Pull request closed without merging**